src/conv-core/isomalloc.C

```diff
@@ -664,233 +728,132 @@ static void CmiIsomallocInitExtent()
-      }
+
       close(fd);
-      if (ss < s || ee > e)
-        CmiAbort(
-            "isomalloc_sync failed during restart, virtual memory regions do not "
-            "overlap.");
+      if (previous.s < IsoRegion.s || previous.e > IsoRegion.e)
+        CmiError("Isomalloc> Warning: virtual memory regions do not overlap: "
+                 "current %" PRIx64 " - %" PRIx64 ", previous %" PRIx64 " - %" PRIx64 "\n",
+                 IsoRegion.s, IsoRegion.e, previous.s, previous.e);
       else
-      {
-        freeRegion.start = (uint8_t *)ss;
-        freeRegion.len = (uint8_t *)ee - (uint8_t *)ss;
-      }
-      CmiPrintf(
-          "Charm++> Consolidated Isomalloc memory region at restart: %p - %p (%d MB).\n",
-          freeRegion.start, freeRegion.start + freeRegion.len, freeRegion.len / meg);
-      goto AFTER_SYNC;
+        IsoRegion = previous;
+
+      DEBUG_PRINT("Charm++> Consolidated Isomalloc memory region at restart: %p - %p (%" PRId64 " MB).\n",
+                  (void *)IsoRegion.s, (void *)IsoRegion.e, (IsoRegion.e - IsoRegion.s) / meg);
     }
+  }
+  else
 #endif
-    if (CmiMyRank() == 0 && freeRegion.len > 0u)
+  if (nosync)
+  {
+    if (CmiMyPe() == 0)
+      CmiPrintf("Isomalloc> Disabling global synchronization of address space.\n");
+  }
+  else if (CmiNumNodes() > 1)
+  {
+    if (CmiMyRank() == 0)
     {
-      if (CmiBarrier() == -1 && CmiMyPe() == 0)
-        CmiAbort("Charm++ Error> +isomalloc_sync requires CmiBarrier() implemented.\n");
-      else
+      auto msg = (CmiAddressSpaceRegionMsg *)CmiAlloc(sizeof(CmiAddressSpaceRegionMsg));
+
+      if (CmiMyNode() == 0)
       {
-        uintptr_t s = (uintptr_t)freeRegion.start;
-        uintptr_t e = (uintptr_t)freeRegion.start + freeRegion.len;
-        int fd, i;
-        char fname[128];
+        DEBUG_PRINT("Charm++> Synchronizing Isomalloc memory region...\n");
 
-        if (CmiMyNode() == 0)
-          CmiPrintf("Charm++> Synchronizing isomalloc memory region...\n");
+        CmiIsomallocSyncWait();
 
-        sprintf(fname, ".isomalloc.%d", CmiMyNode());
+        DEBUG_PRINT("Isomalloc> Node %d sending region for assignment: %" PRIx64 " %" PRIx64 "\n",
+                    CmiMyNode(), IsoRegion.s, IsoRegion.e);
+        msg->region = IsoRegion;
+        CmiSetHandler((char *)msg, CmiIsomallocSyncHandlerOtherNodesIdx);
+        CmiSyncNodeBroadcastAndFree(sizeof(CmiAddressSpaceRegionMsg), msg);
 
-        /* remove file before writing for safe */
-        unlink(fname);
-#if CMK_HAS_SYNC && !CMK_DISABLE_SYNC
-        if (system("sync") == -1)
-        {
-          CmiAbort(
-              "Isomalloc> call to system(\"sync\") failed while synchronizing memory "
-              "regions!");
-        }
-#endif
+        CsdSchedulePoll();
 
-        CmiBarrier();
+        if (IsoRegion.s >= IsoRegion.e)
+          CmiAbort("Isomalloc> failed to find consolidated region: %" PRIx64 " - %" PRIx64 ".\n", IsoRegion.s, IsoRegion.e);
```

Comment:
Maybe add here that users can run with +no_isomalloc_sync if they don't need it

Reply:
Done